src/stream/src/executor/source/kafka_backfill_executor.rs

```diff
@@ -0,0 +1,888 @@
+                    left,
+                    futures::stream::pending().boxed().map(Either::Right),
+                    select_strategy,
+                );
+                // XXX: do we have to store the original reader? Can we simply rebuild the reader later?
+                paused_reader = Some(right);
+            };
+        }
+
+        // If the first barrier requires us to pause on startup, pause the stream.
+        if barrier.is_pause_on_startup() {
+            pause_reader!();
+        }
+
+        yield Message::Barrier(barrier);
+
+        let mut backfill_stage = BackfillStage {
+            states: backfill_states,
+            unfinished_splits,
+        };
+
+        if !self.backfill_finished(&backfill_stage.states).await? {
+            // We allow data to flow for `WAIT_BARRIER_MULTIPLE_TIMES` * `expected_barrier_latency_ms`
+            // milliseconds, considering some other latencies like network and cost in Meta.
+            let mut max_wait_barrier_time_ms = self.system_params.load().barrier_interval_ms()
+                as u128
+                * WAIT_BARRIER_MULTIPLE_TIMES;
+            let mut last_barrier_time = Instant::now();
+            let mut self_paused = false;
+
+            'backfill_loop: while let Some(either) = backfill_stream.next().await {
+                match either {
+                    // Upstream
+                    Either::Left(msg) => {
+                        let Ok(msg) = msg else {
+                            let e = msg.unwrap_err();
+                            let core = &self.stream_source_core;
+                            tracing::warn!(
+                                error = ?e.as_report(),
+                                actor_id = self.actor_ctx.id,
+                                source_id = %core.source_id,
+                                "stream source reader error",
+                            );
+                            GLOBAL_ERROR_METRICS.user_source_reader_error.report([
+                                "SourceReaderError".to_owned(),
+                                e.to_report_string(),
+                                "SourceExecutor".to_owned(),
+                                self.actor_ctx.id.to_string(),
+                                core.source_id.to_string(),
+                            ]);
+
+                            let reader = self
+                                .build_stream_source_reader(
+                                    &source_desc,
+                                    backfill_stage.unfinished_splits.clone(),
+                                )
+                                .await?;
+
+                            backfill_stream = select_with_strategy(
+                                input.by_ref().map(Either::Left),
+                                reader.map(Either::Right),
+                                select_strategy,
+                            );
+                            continue;
```

Comment:
Any reason why we don't just throw this error? Do we want to retry from the last kafka checkpoint offset?

Reply:
In most cases, the error is a timeout or something related to network jitter. For example, the upstream broker is down, let it retry inside source instead of letting the cluster fall into a crash loop is a better choice.

Reply:
But it looks like it would emit some duplicate rows to the downstream, if we don't clean up the downstream state.🤔

Reply:
Here we build readers from the latest offsets. (`unfinished_splits` contains latest offsets)

Reply:
Really? Doesn't `states` of `BackfillStage` contain the latest offsets?

Reply:
> Really? Doesn't `states` of `BackfillStage` contain the latest offsets?

You are right. This should be a bug. The logic should be the same as `if split_changed { ...` 🤡

Reply:
fixed